app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java

```diff
@@ -0,0 +1,40 @@
+package it.niedermann.owncloud.notes.persistence.migration;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+import androidx.annotation.NonNull;
+import androidx.preference.PreferenceManager;
+import androidx.room.migration.Migration;
+import androidx.sqlite.db.SupportSQLiteDatabase;
+
+/**
+ * Enabling backgroundSync, set from {@link String} values to {@link Boolean} values
+ * https://github.com/stefan-niedermann/nextcloud-notes/issues/1168
+ */
+
+public class Migration_21_22 extends Migration {
+    @NonNull
+    private final Context context;
+
+    public Migration_21_22(@NonNull Context context) {
+        super(21, 22);
+        this.context = context;
+    }
+
+    @Override
+    public void migrate(@NonNull SupportSQLiteDatabase database) {
+        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        if (sharedPreferences.contains("backgroundSync")) {
+            if (sharedPreferences.getString("backgroundSync", "on").equals("off")) {
+                editor.remove("backgroundSync");
+                editor.putBoolean("backgroundSync", false);
+            } else {
+                editor.remove("backgroundSync");
+                editor.putBoolean("backgroundSync", true);
+            }
+        }
+        editor.apply();
```

Comment:
We need to update the `SyncWorker` to make it aware of the "new" setting:

```java
SyncWorker.update(context, true);
```

I'd say we need to call it with `true` in case the `SharedPreferences`
- did **not** contain a setting, because the default behavior now changed
- **did** contain a setting which was **not** `off`

Otherwise the new default will only be applied on new installations and installations which change the setting after performing the update.

I recommend you to test the behavior with the various combinations (No setting, setting `off`, setting `15_minutes`, setting `6_hours`, etc...) with this approach:

1. Install the Notes app on an Android device and set it up
2. Make a snapshot of the emulator (three dots menu -> Take snapshot)
  ![grafik](https://user-images.githubusercontent.com/4741199/116893719-95b56480-ac31-11eb-9e46-99b6be9b6149.png)
3. Validate the old setting in the device explorer of Android Studio
  ![grafik](https://user-images.githubusercontent.com/4741199/116893834-b67dba00-ac31-11eb-95d1-8b98c2a03db5.png)
4. Install the branch version
5. Validate the new setting in the device explorer (Make sure to right click -> synchronize first, before validating)
6. Check the Logcat (filter for `SyncWorker`) to verify whether it run or not
7. Repeat for each combination (you can now easily restart from your snapshot, see point 2)

Reply:
Is it that we need to call `SyncWorker.update` during `Migration`, otherwise the settings are not updated?

Reply:
Yes, during the migration.

Oh, and i just saw, that we also need to call it additionally for new installations. So also call it when importing an account the first time:

https://github.com/stefan-niedermann/nextcloud-notes/blob/0348f0a642df5d1a35008f7adcd95de1f4449daf/app/src/main/java/it/niedermann/owncloud/notes/importaccount/ImportAccountActivity.java#L91 (right before `runOnUiThread(() -> {...`) with `true`.

Reply:
I have updated in migration, and in `ImportAccountActivity.java`, I dont know whether my operation is correct or not.